README.md
# ts-graylog

TypeScript Graylog client library for Node.js, based on [node-graylog2](http://github.com/Wizcorp/node-graylog2). This has been heavily modified to the point where there is not much left of the original.

It is a rewrite using TypeScript.


CHANGELOG
1.0.3: October 24, 2018
* improvement on JSON stringify to exclude strings and numbers.
1.0.2: October 23, 2018
* add number as an addl type for the keys `value` additionalFields. Moved inline definition to a `type` for the additionalField definition.
1.0.1: October 17, 2018
* fix to parsing objects of type Error. Created a Helper for passing into the JSON.stringify.
1.0.0: October 12, 2018
* initial release


tests/index.spec.ts
import { Graylog } from "../src";
import { GraylogMessage } from "../src/models/message.model";
let graylog: Graylog;

/**
 * Since Graylog communication is UDP there is not real way to confirm
 * a message was received, besides making sure is not throwing an error
 */
describe("Graylog object tests", () => {

    beforeEach(() => {
        graylog = new Graylog();
    });

    test("graylog instance should not be null", () => {
        expect(graylog).not.toBeNull();
    });

    it("should have a valid config file", () => {
        expect(graylog.config).not.toBeNull();
    });

    it("should have at least one server", () => {
        expect(graylog.config.servers).not.toBeNull();
    });

    it("should send an emergency log", () => {
        const message = new GraylogMessage({
            short_message: "Sup!",
            full_message: "What's it is up?"
        });
        const error = new Error("Hi, I am an error");
        expect(graylog.emergency(message, error)).toBeUndefined();
    });

    it("should send an info log", () => {
        const message = new GraylogMessage({
            short_message: "Sup!",
            full_message: "What's it is up?"
        });
        expect(graylog.info(message)).toBeUndefined();
    });

    it("should send an debug log with additional info", () => {
        const message = new GraylogMessage({
            short_message: "Sup!",
            full_message: "What's it is up?"
        });
        const error = new Error("Hi, I am an error");
        expect(graylog.debug(message, error, { _test: "hi, I am an additional field" })).toBeUndefined();
    });
});

tests/config.spec.ts
import os from "os";
import { GraylogConfig, IGraylogConfig } from "../src/models/config.model";
let graylogConfig: IGraylogConfig;
const hostname: string = os.hostname();
const bufferSize = 1400;

describe("Graylog Config object tests", () => {

    beforeEach(() => {
        graylogConfig = new GraylogConfig();
    });

    test("graylog config instance should not be null", () => {
        expect(graylogConfig).not.toBeNull();
    });

    it("should have the minimum valid servers list", () => {
        expect(graylogConfig.servers).not.toBeNull();
    });

    it("should have the default deflate option", () => {
        expect(graylogConfig.deflate).toBe("optimal");
    });

    it("should have the default hostname", () => {
        expect(graylogConfig.hostname).toBe(hostname);
    });

    it("should have the default bufferSize", () => {
        expect(graylogConfig.bufferSize).toBe(bufferSize);
    });
});

tests/message.spec.ts
import { IGraylogMessage, GraylogMessage } from "../src/models/message.model";
let graylogMessage: IGraylogMessage;
const testShortMessage = "Short Message";
const testFullMessage = "Full Message";
const testRandomField = "Hi There!";

describe("Graylog Message object tests", () => {

    beforeEach(() => {
        graylogMessage = new GraylogMessage({
            short_message: testShortMessage,
            full_message: testFullMessage,
            test_random_field: testRandomField
        });
    });

    test("graylog message instance should not be null", () => {
        expect(graylogMessage).not.toBeNull();
    });

    it("should have the short message", () => {
        expect(graylogMessage.short_message).toBe(testShortMessage);
    });

    it("should have the full message", () => {
        expect(graylogMessage.full_message).toBe(testFullMessage);
    });

    it("should have a random field", () => {
        expect(graylogMessage.test_random_field).toBe(testRandomField);
    });
});

src/index.ts
/**
 * Node modules
 */
import { deflate } from "zlib";
import { randomBytes } from "crypto";
import { createSocket, Socket } from "dgram";
import { EventEmitter } from "events";

/**
 * Local modules
 */
import { IGraylogConfig, GraylogConfig, IGraylogConfigServer, GraylogConfigServer } from "./models/config.model";
import { GraylogLevelEnum } from "./models/enums.model";
import { GraylogMessage, IGraylogMessage } from "./models/message.model";
import { ReplacerHelper } from "./helpers/replacer.helper";

type AdditionalFields = { [key: string]: string | number | undefined };

/**
 * Graylog main class
 */
export class Graylog extends EventEmitter {
    // Private
    private _callCount: number;
    private _isDestroyed: boolean;
    private _client: Socket;

    // Public
    readonly config: IGraylogConfig;

    constructor(config?: IGraylogConfig) {
        super();

        if (!config) {
            config = {};
        }

        this.config = new GraylogConfig(config);
        this._callCount = 0;

        this._isDestroyed = false;
        this._client = createSocket("udp4");
    }

    private getClient(): Socket {
        if (!this._client && !this._isDestroyed) {
            this._client = createSocket("udp4");

            this._client.on("error", (err) => {
                this.emit("error", err);
            });
        }

        return this._client;
    }

    private getServer(): IGraylogConfigServer {
        if (this.config.servers) {
            return this.config.servers[this._callCount++ % this.config.servers.length];
        }
        return new GraylogConfigServer();
    }

    destroy() {
        if (this._client) {
            this._client.close();
            this._client.removeAllListeners();
            this._isDestroyed = true;
        }
    }

    emergency(message: IGraylogMessage, error: Error, additionalFields?: AdditionalFields, timestamp?: number) {
        return this._log(message, error, additionalFields, GraylogLevelEnum.EMERG, timestamp);
    }

    alert(message: IGraylogMessage, error: Error, additionalFields?: AdditionalFields, timestamp?: number) {
        return this._log(message, error, additionalFields, GraylogLevelEnum.ALERT, timestamp);
    }

    critical(message: IGraylogMessage, error: Error, additionalFields?: AdditionalFields, timestamp?: number) {
        return this._log(message, error, additionalFields, GraylogLevelEnum.CRIT, timestamp);
    }

    error(message: IGraylogMessage, error: Error, additionalFields?: AdditionalFields, timestamp?: number) {
        return this._log(message, error, additionalFields, GraylogLevelEnum.ERROR, timestamp);
    }

    warning(message: IGraylogMessage, error: Error, additionalFields?: AdditionalFields, timestamp?: number) {
        return this._log(message, error, additionalFields, GraylogLevelEnum.WARNING, timestamp);
    }

    notice(message: IGraylogMessage, additionalFields?: AdditionalFields, timestamp?: number) {
        return this._log(message, undefined, additionalFields, GraylogLevelEnum.NOTICE, timestamp);
    }

    info(message: IGraylogMessage, additionalFields?: AdditionalFields, timestamp?: number) {
        return this._log(message, undefined, additionalFields, GraylogLevelEnum.INFO, timestamp);
    }

    debug(message: IGraylogMessage, error?: Error, additionalFields?: AdditionalFields, timestamp?: number) {
        return this._log(message, error, additionalFields, GraylogLevelEnum.DEBUG, timestamp);
    }

    log(message: IGraylogMessage, error?: Error, additionalFields?: AdditionalFields) {
        return this._log(message, error, additionalFields, message.level);
    }

    private _log(message: IGraylogMessage, error?: Error, additionalFields?: AdditionalFields, level?: GraylogLevelEnum, timestamp?: number) {
        let payload;

        if (message) {
            if (!(message instanceof GraylogMessage)) {
                message = new GraylogMessage(message);
            }
        } else {
            message = new GraylogMessage({
                timestamp: timestamp,
                level: level
            });
        }

        if (error) {
            message.exception = JSON.stringify(error, ReplacerHelper.replaceError);
        }

        // We insert additional fields
        if (additionalFields) {
            for (const key in additionalFields) {
                if (additionalFields.hasOwnProperty(key)) {
                    const element = additionalFields[key];
                    if (element === undefined) {
                        continue;
                    }

                    if (typeof element === "string" || typeof element === "number") {
                        message["_" + key] = element;
                    } else {
                        message["_" + key] = JSON.stringify(element);
                    }
                }
            }
        }

        // https://github.com/Graylog2/graylog2-docs/wiki/GELF
        if (message._id) {
            message.__id = message._id;
            delete message._id;
        }

        // Compression
        payload = new Buffer(JSON.stringify(message));

        if (this.config.deflate === "never" || (this.config.bufferSize && this.config.deflate === "optimal" && payload.length <= this.config.bufferSize)) {
            this.sendPayload(undefined, payload);
        } else {
            deflate(payload, this.sendPayload.bind(this));
        }
    }

    private sendPayload(err: any, buffer: Buffer) {
        if (err) {
            return this.emitError(err);
        }

        // If it all fits, just send it
        if (this.config.bufferSize && buffer.length <= this.config.bufferSize) {
            const server = this.getServer();
            return this.send(buffer, server, undefined);
        }

        // It didn't fit, so prepare for a chunked stream
        if (this.config.bufferSize) {
            const bufferSize = this.config.bufferSize;
            const dataSize = bufferSize - 12;  // the data part of the buffer is the buffer size - header size
            const chunkCount = Math.ceil(buffer.length / dataSize);

            if (chunkCount > 128) {
                return this.emitError("Cannot log messages bigger than " + (dataSize * 128) + " bytes");
            }

            // Generate a random id in buffer format
            randomBytes(8, (err, id) => {
                if (err) {
                    return this.emitError(err);
                }

                // To be tested: what's faster, sending as we go or prebuffering?
                const server = this.getServer();
                const chunk = new Buffer(bufferSize);
                let chunkSequenceNumber = 0;

                // Prepare the header

                // Set up magic number (bytes 0 and 1)
                chunk[0] = 30;
                chunk[1] = 15;

                // Set the total number of chunks (byte 11)
                chunk[11] = chunkCount;

                // Set message id (bytes 2-9)
                id.copy(chunk, 2, 0, 8);

                const send = (err?: any) => {
                    if (err || chunkSequenceNumber >= chunkCount) {
                        // We have reached the end, or had an error (which will already have been emitted)
                        return;
                    }

                    // Set chunk sequence number (byte 10)
                    chunk[10] = chunkSequenceNumber;

                    // Copy data from full buffer into the chunk
                    const start = chunkSequenceNumber * dataSize;
                    const stop = Math.min((chunkSequenceNumber + 1) * dataSize, buffer.length);

                    buffer.copy(chunk, 12, start, stop);

                    chunkSequenceNumber++;

                    // Send the chunk
                    this.send(chunk.slice(0, stop - start + 12), server, send);
                };

                send();
            });
        }
    }

    private send(chunk: Buffer | string | Uint8Array, server: IGraylogConfigServer, cb?: (error: Error | null, bytes?: number) => void) {
        const client = this.getClient();

        if (!client) {
            const error = new Error("Socket was already destroyed");

            this.emit("error", error);
            if (cb) {
                return cb(error);
            }
            return;
        }

        client.send(chunk, 0, chunk.length, server.port, server.host, (err) => {
            if (err) {
                this.emit("error", err);
            }

            if (cb) {
                cb(err);
            }
        });
    }

    private emitError(err: any) {
        this.emit("error", err);
        console.error(err);
    }
}


jest.config.js
// For a detailed explanation regarding each configuration property, visit:
// https://jestjs.io/docs/en/configuration.html

module.exports = {
  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after the first failure
  // bail: false,

  // Respect "browser" field in package.json when resolving modules
  // browser: false,

  // The directory where Jest should store its cached dependency information
  // cacheDirectory: "/var/folders/1_/hnytw6j54lj6pf2585rf7q9w0000gn/T/jest_dx",

  // Automatically clear mock calls and instances between every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  // collectCoverage: false,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: null,

  // The directory where Jest should output its coverage files
  coverageDirectory: "coverage",

  // An array of regexp pattern strings used to skip coverage collection
  // coveragePathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // A list of reporter names that Jest uses when writing coverage reports
  // coverageReporters: [
  //   "json",
  //   "text",
  //   "lcov",
  //   "clover"
  // ],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: null,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // Force coverage collection from ignored files usin a array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: null,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: null,

  // A set of global variables that need to be available in all test environments
  globals: {
    "ts-jest": {
      "tsConfig": "tsconfig.json"
    }
  },

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   "node_modules"
  // ],

  // An array of file extensions your modules use
  moduleFileExtensions: [
    "ts",
    "tsx",
    "js"
  ],

  // A map from regular expressions to module names that allow to stub out resources with a single module
  // moduleNameMapper: {},

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  // modulePathIgnorePatterns: [],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: "always",

  // A preset that is used as a base for Jest's configuration
  // preset: null,

  // Run tests from one or more projects
  // projects: null,

  // Use this configuration option to add custom reporters to Jest
  // reporters: undefined,

  // Automatically reset mock state between every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: null,

  // Automatically restore mock state between every test
  // restoreMocks: false,

  // The root directory that Jest should scan for tests and modules within
  rootDir: "./tests",

  // A list of paths to directories that Jest should use to search for files in
  // roots: [
  //   "<rootDir>"
  // ],

  // Allows you to use a custom runner instead of Jest's default test runner
  // runner: "jest-runner",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  // setupFiles: [],

  // The path to a module that runs some code to configure or set up the testing framework before each test
  // setupTestFrameworkScriptFile: null,

  // A list of paths to snapshot serializer modules Jest should use for snapshot testing
  // snapshotSerializers: [],

  // The test environment that will be used for testing
  testEnvironment: "node",

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns Jest uses to detect test files
  testMatch: [
    "**/tests/*.spec.+(ts|tsx|js)"
  ],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  // testPathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // The regexp pattern Jest uses to detect test files
  // testRegex: "",

  // This option allows the use of a custom results processor
  // testResultsProcessor: null,

  // This option allows use of a custom test runner
  // testRunner: "jasmine2",

  // This option sets the URL for the jsdom environment. It is reflected in properties such as location.href
  // testURL: "http://localhost",

  // Setting this value to "fake" allows the use of fake timers for functions such as "setTimeout"
  // timers: "real",

  // A map from regular expressions to paths to transformers
  transform: {
    "^.+\\.(ts|tsx)$": "ts-jest"
  },

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  // transformIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  // verbose: null,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
};


src/models/config.model.ts
import os from "os";

/**
 * Interface to provide the config object required by the main
 * Graylog object to start working
 */
export interface IGraylogConfig {
    readonly servers?: Array<IGraylogConfigServer>;
    readonly hostname?: string;
    readonly deflate?: "optimal" | "always" | "never";
    readonly bufferSize?: number;
}

/**
 * Implementation for the @type {IGraylogConfig} interface.
 */
export class GraylogConfig implements IGraylogConfig {
    private DEFAULT_BUFFERSIZE = 1400;

    readonly servers: Array<IGraylogConfigServer>;
    readonly hostname: string;
    readonly deflate: "optimal" | "always" | "never";
    readonly bufferSize: number;

    constructor(data?: IGraylogConfig) {
        if (!data) {
            data = {} as IGraylogConfig;
        }

        this.servers = new Array<IGraylogConfigServer>();
        if (data.servers) {
            data.servers.forEach(server => {
                this.servers.push(new GraylogConfigServer(server));
            });
        } else {
            this.servers.push(new GraylogConfigServer());
        }

        this.hostname = data.hostname || os.hostname();
        this.deflate = data.deflate || "optimal";
        this.bufferSize = data.bufferSize || this.DEFAULT_BUFFERSIZE;
    }
}

/**
 * Interface to provide the required properties for the server object
 */
export interface IGraylogConfigServer {
    readonly host: string;
    readonly port: number;
}

/**
 * Implementation for the @type {IGraylogConfigServer} interface.
 */
export class GraylogConfigServer implements IGraylogConfigServer {
    private DEFAULT_HOST = "127.0.0.1";
    private DEFAULT_PORT = 12201;

    readonly host: string;
    readonly port: number;

    constructor(data?: IGraylogConfigServer) {
        if (!data) {
            data = {} as IGraylogConfigServer;
        }

        this.host = data.host || this.DEFAULT_HOST;
        this.port = data.port || this.DEFAULT_PORT;
    }
}


src/models/message.model.ts
/**
 * Node modules
 */
import os from "os";
import { GraylogLevelEnum } from "./enums.model";

/**
 * Interface to send messages to Graylog. It contains the basic set of properties
 * but also the ability to have more properties on the fly.
 */
export interface IGraylogMessage {
    [key: string]: any; // for properties on the fly
    short_message?: string;
    full_message?: string;
    host?: string;
    version?: string;
    level?: GraylogLevelEnum;
    exception?: string;
    readonly timestamp?: number;
}

/**
 * Implementation for the @type {IGraylogMessage}
 */
export class GraylogMessage implements IGraylogMessage {
    [key: string]: any;
    short_message?: string;
    full_message?: string;
    version?: string;
    host?: string;
    level?: GraylogLevelEnum;
    exception?: string;
    readonly timestamp: number;

    constructor(data?: IGraylogMessage) {
        if (!data) {
            data = {} as IGraylogMessage;
        }

        Object.keys(data).forEach(d => {
            if (data) {
                this[d] = data[d];
            }
        });

        this.version = data.version || "1.1";
        this.timestamp = data.timestamp || new Date().getTime() / 1000;
        this.host = data.host || os.hostname();
        this.level = data.level || GraylogLevelEnum.INFO;
        this.exception = data.exception;
    }
}

graylog.js
import { deflate } from 'zlib';
import { randomBytes } from 'crypto';
import { createSocket } from 'dgram';
import { inherits } from 'util';
import { EventEmitter } from 'events';
import assert from 'assert';

/**
 * Graylog instances emit errors. That means you really really should listen for them,
 * or accept uncaught exceptions (node throws if you don't listen for "error").
 */

var graylog = function graylog(config) {
    EventEmitter.call(this);

    this.config       = config;

    this.servers      = config.servers;
    this.client       = null;
    this.hostname     = config.hostname || require('os').hostname();
    this.facility     = config.facility || 'Node.js';
    this.deflate      = config.deflate || 'optimal';
    assert(
      this.deflate === 'optimal' || this.deflate === 'always' || this.deflate === 'never',
      'deflate must be one of "optimal", "always", or "never". was "' + this.deflate + '"');

    this._unsentMessages = 0;
    this._unsentChunks = 0;
    this._callCount   = 0;

    this._onClose = null;
    this._isDestroyed = false;

    this._bufferSize  = config.bufferSize || this.DEFAULT_BUFFERSIZE;
};

inherits(graylog, EventEmitter);

graylog.prototype.DEFAULT_BUFFERSIZE = 1400;  // a bit less than a typical MTU of 1500 to be on the safe side

graylog.prototype.level = {
    EMERG: 0,    // system is unusable
    ALERT: 1,    // action must be taken immediately
    CRIT: 2,     // critical conditions
    ERR: 3,      // error conditions
    ERROR: 3,    // because people WILL typo
    WARNING: 4,  // warning conditions
    NOTICE: 5,   // normal, but significant, condition
    INFO: 6,     // informational message
    DEBUG: 7     // debug level message
};

graylog.prototype.getServer = function () {
    return this.servers[this._callCount++ % this.servers.length];
};

graylog.prototype.getClient = function () {
    if (!this.client && !this._isDestroyed) {
        this.client = createSocket("udp4");

        var that = this;
        this.client.on('error', function (err) {
            that.emit('error', err);
        });
    }

    return this.client;
};

graylog.prototype.destroy = function () {
    if (this.client) {
        this.client.close();
        this.client.removeAllListeners();
        this.client = null;
        this._onClose = null;
        this._isDestroyed = true;
    }
};

graylog.prototype.emergency = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.EMERG);
};

graylog.prototype.alert = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.ALERT);
};

graylog.prototype.critical = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.CRIT);
};

graylog.prototype.error = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.ERROR);
};

graylog.prototype.warning = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.WARNING);
};
graylog.prototype.warn = graylog.prototype.warning;

graylog.prototype.notice = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.NOTICE);
};

graylog.prototype.info = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.INFO);
};
graylog.prototype.log = graylog.prototype.info;

graylog.prototype.debug = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.DEBUG);
};

graylog.prototype._log = function log(short_message, full_message, additionalFields, timestamp, level) {
    this._unsentMessages += 1;

    var payload,
        fileinfo,
        that    = this,
        field   = '',
        message = {
            version    : '1.1',
            timestamp  : (timestamp || new Date()).getTime() / 1000,
            host       : this.hostname,
            facility   : this.facility,
            level      : level
        };

    if (typeof(short_message) !== 'object' && typeof(full_message) === 'object' && additionalFields === undefined) {
        // Only short message and additional fields are available
        message.short_message   = short_message;
        message.full_message    = short_message;

        additionalFields = full_message;
    } else  if (typeof(short_message) !== 'object') {
        // We normally set the data
        message.short_message   = short_message;
        message.full_message    = full_message || short_message;
    } else if (short_message.stack && short_message.message) {

        // Short message is an Error message, we process accordingly
        message.short_message = short_message.message;
        message.full_message  = short_message.stack;

        // extract error file and line
        fileinfo = message.stack.split('\n')[0];
        fileinfo = fileinfo.substr(fileinfo.indexOf('('), fileinfo.indexOf(')'));
        fileinfo = fileinfo.split(':');

        message.file = fileinfo[0];
        message.line = fileinfo[1];

        additionalFields = full_message || additionalFields;
    } else {
        message.full_message = message.short_message = JSON.stringify(short_message);
    }

    // We insert additional fields
    for (field in additionalFields) {
        message['_' + field] = additionalFields[field];
    }

    // https://github.com/Graylog2/graylog2-docs/wiki/GELF
    if (message._id) {
        message.__id = message._id;
        delete message._id;
    }

    // Compression
    payload = new Buffer(JSON.stringify(message));

    function sendPayload(err, buffer) {
        if (err) {
            that._unsentMessages -= 1;
            return that.emitError(err);
        }

        // If it all fits, just send it
        if (buffer.length <= that._bufferSize) {
            that._unsentMessages -= 1;
            return that.send(buffer, that.getServer());
        }

        // It didn't fit, so prepare for a chunked stream

        var bufferSize = that._bufferSize;
        var dataSize   = bufferSize - 12;  // the data part of the buffer is the buffer size - header size
        var chunkCount = Math.ceil(buffer.length / dataSize);

        if (chunkCount > 128) {
            that._unsentMessages -= 1;
            return that.emitError('Cannot log messages bigger than ' + (dataSize * 128) +  ' bytes');
        }

        // Generate a random id in buffer format
        randomBytes(8, function (err, id) {
            if (err) {
                that._unsentMessages -= 1;
                return that.emitError(err);
            }

            // To be tested: what's faster, sending as we go or prebuffering?
            var server = that.getServer();
            var chunk = new Buffer(bufferSize);
            var chunkSequenceNumber = 0;

            // Prepare the header

            // Set up magic number (bytes 0 and 1)
            chunk[0] = 30;
            chunk[1] = 15;

            // Set the total number of chunks (byte 11)
            chunk[11] = chunkCount;

            // Set message id (bytes 2-9)
            id.copy(chunk, 2, 0, 8);

            function send(err) {
                if (err || chunkSequenceNumber >= chunkCount) {
                    // We have reached the end, or had an error (which will already have been emitted)
                    that._unsentMessages -= 1;
                    return;
                }

                // Set chunk sequence number (byte 10)
                chunk[10] = chunkSequenceNumber;

                // Copy data from full buffer into the chunk
                var start = chunkSequenceNumber * dataSize;
                var stop  = Math.min((chunkSequenceNumber + 1) * dataSize, buffer.length);

                buffer.copy(chunk, 12, start, stop);

                chunkSequenceNumber++;

                // Send the chunk
                that.send(chunk.slice(0, stop - start + 12), server, send);
            }

            send();
        });
    }

    if (this.deflate === 'never' || (this.deflate === 'optimal' && payload.length <= this._bufferSize)) {
      sendPayload(null, payload);
    } else {
      deflate(payload, sendPayload);
    }
};

graylog.prototype.send = function (chunk, server, cb) {
    var that = this,
        client = this.getClient();

    if (!client) {
        var error = new Error('Socket was already destroyed');

        this.emit('error', error);
        return cb(error);
    }

    this._unsentChunks += 1;

    client.send(chunk, 0, chunk.length, server.port, server.host, function (err) {
        that._unsentChunks -= 1;

        if (err) {
            that.emit('error', err);
        }

        if (cb) {
            cb(err);
        }

        if (that._unsentChunks === 0 && that._unsentMessages === 0 && that._onClose) {
            that._onClose();
        }
    });
};

graylog.prototype.emitError = function (err) {
    this.emit('error', err);

    if (this._unsentChunks === 0 && this._unsentMessages === 0 && this._onClose) {
        this._onClose();
    }
};

graylog.prototype.close = function (cb) {
    var that = this;

    if (this._onClose || this._isDestroyed) {
        return process.nextTick(function () {
            var error = new Error('Close was already called once');

            if (cb) {
                return cb(error);
            }

            that.emit('error', error);
        });
    }

    this._onClose = function () {
        that.destroy();

        if (cb) {
            cb();
        }
    };

    if (this._unsentChunks === 0 && this._unsentMessages === 0) {
        process.nextTick(function () {
            that._onClose();
        });
    }
};

const _graylog = graylog;
export { _graylog as graylog };
